File: 5_semestre/Programacao-para-Dispositivos-Moveis/UAS/UA1/csharp/Empresa/Desenvolvedor.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Empresa;

public class Desenvolvedor : Funcionario
{
    public NivelExperiencia Nivel       { get; private set; }
    public List<string>     Linguagens  { get; private set; }

    public Desenvolvedor(string nome, string cpf, double salarioBase, Departamento departamento,
                        DateTime dataAdmissao, NivelExperiencia nivel, List<string> linguagens)
        : base(nome, cpf, salarioBase, departamento, dataAdmissao)
    {
        Nivel = nivel;
        Linguagens = new List<string>(linguagens);
    }

    public void AdicionarLinguagem(string linguagem)
    {
        if (!Linguagens.Contains(linguagem))
        {
            Linguagens.Add(linguagem);
        }
    }

    public override double CalcularBonus()
    {
        return Nivel switch
        {
            NivelExperiencia.Estagiario     => SalarioBase * 0.05,
            NivelExperiencia.Junior         => SalarioBase * 0.08,
            NivelExperiencia.Pleno          => SalarioBase * 0.12,
            NivelExperiencia.Senior         => SalarioBase * 0.18,
            NivelExperiencia.Especialista   => SalarioBase * 0.25,
            _ => 0
        };
    }

    public override double CalcularSalarioTotal() => SalarioBase + CalcularBonus();

    public override string ToString()
    {
        return $"[DEV] {base.ToString()} | Nível: {Nivel} | Linguagens: {string.Join(", ", Linguagens)}";
    }
}

File: 5_semestre/Programacao-para-Dispositivos-Moveis/UAS/UA1/dart/empresa/desenvolvedor.dart
class Desenvolvedor extends Funcionario {
  NivelExperiencia nivel;
  List<String> linguagens;

  Desenvolvedor({
    required String nome,
    required String cpf,
    required double salarioBase,
    required Departamento departamento,
    required DateTime dataAdmissao,
    required this.nivel,
    required List<String> linguagens,
  })  : linguagens = List.from(linguagens),
        super(
          nome: nome,
          cpf: cpf,
          salarioBase: salarioBase,
          departamento: departamento,
          dataAdmissao: dataAdmissao,
        );

  void adicionarLinguagem(String linguagem) {
    if (!linguagens.contains(linguagem)) {
      linguagens.add(linguagem);
    }
  }

  @override
  double calcularBonus() {
    switch (nivel) {
      case NivelExperiencia.ESTAGIARIO:
        return salarioBase * 0.05;
      case NivelExperiencia.JUNIOR:
        return salarioBase * 0.08;
      case NivelExperiencia.PLENO:
        return salarioBase * 0.12;
      case NivelExperiencia.SENIOR:
        return salarioBase * 0.18;
      case NivelExperiencia.ESPECIALISTA:
        return salarioBase * 0.25;
    }
  }

  @override
  double calcularSalarioTotal() {
    return salarioBase + calcularBonus();
  }

  @override
  String toString() {
    return "[DEV] ${super.toString()} | Nível: $nivel | Linguagens: ${linguagens.join(", ")}";
  }
}